0006-scriptless-scripts-with-ecdsa.adoc

```diff
@@ -0,0 +1,79 @@
+
+First P2 sends g^a^ to P1.
+It is important to do this first to avoid breaking the simulation by letting P2 choose g^α^ adaptively.
+
+After that, P1 and P2 have exchange R~1~ = g^r~1~^ and R~2~ = g^r~2~^ via the usual simulatable Diffie-Hellman coin flipping in <<Lindell17>>.
+
+Next, P2 sends R~3~ = R~2~^α^ and a proof that (g^α^, R~2~, R~3~) is a valid Diffie-Hellman triple.
+Note, I think that he could actually just send this when he sends R~2~ during the coin flipping but I'm unsure whether the two things can happen in parallel without breaking the simulation.
+
+The proof for a valid Diffie-Hellman triple used in <<kzen_multi_hop_locks>> works by proving knowledge of a pre-image α for the one-way homomorphism h(α) = (g^α^, R~2~^α^) through the Sigma protocol.
+This proves that (g^α^, R~2~, R~3~) is a valid Diffie-Hellman triple.
+I found a nice explanation of this in <<Maurer09>> (section 6.6).
+
+=== Implementation Notes
+
+The proof is implemented here: https://github.com/KZen-networks/curv/blob/18f0081a9a3025eafd49ad496e5032647edf4aa3/src/cryptographic_primitives/proofs/sigma_ec_ddh.rs
+and is non-interactive via Fiat-Shamir transform of the Sigma protocol.
+
+== Signing
+
+The signing protocol then continues as normal under R except at the end the `s` value for the signature that P1 has is not yet valid (it's missing α).
+We'll call this incomplete `s` value s~1~ (it's s'' in the paper)
+
+P1 now sends s~1~ to P2.
+P2 can now finally calculate the final s from s~1~ using his knowledge of α.
+Specifically s = α^-1^(s~1~).
+
+When s is revealed to P1, she can extract α = (s(s~1~^-1^))^-1^.
+
+Well almost.
+The (R,s) signature that makes it into the Bitcoin or Ethereum blockhain isn't necessarily the one you will be able to extract α from directly.
+This is because (R, -s) is also a valid signature.
+This isn't mentioned in <<Sanchez18a>> but it is recognised in <<Sanchez18b>>.
+You will have to try both α and -α to see which one is the real α.
+
+=== Implementation Notes
+
+This will have to implemented by us, but it may be helpful to follow along with：https://github.com/KZen-networks/multi-hop-locks to see how they did it for the multi-hop lock case.
+
+[bibliography]
+== References
+
+- [[[Lindell17]]] Fast Secure Two-Party ECDSA Signing: https://eprint.iacr.org/2017/552.pdf
+- [[[Sanchez18a]]] Scriptless Scripts with ECDSA: https://lists.linuxfoundation.org/pipermail/lightning-dev/attachments/20180426/fe978423/attachment-0001.pdf
+- [[[Sanchez18b]]] Anonymous Multi-Hop Locks for Blockchain Scalability and Interoperability: https://eprint.iacr.org/2018/472.pdf
+- [[[kzen_multi_hop_locks]]] KZen's ECDSA multi hop locks implementation https://github.com/KZen-networks/multi-hop-locks
+- [[[Maurer09]]] Unifying Zero-Knowledge Proofs of Knowledge：ftp://ftp.inf.ethz.ch/pub/crypto/publications/Maurer09.pdf
```

Comment:
:wrench: Broken link, does this fix it?
```suggestion
- [[[Maurer09]]] Unifying Zero-Knowledge Proofs of Knowledge ftp://ftp.inf.ethz.ch/pub/crypto/publications/Maurer09.pdf
```

Reply:
I put some weird math colon "：" instead of ":" with this `xah-math-input-mode` thing on emacs.